Issue #3102: DesignForExtension: consider overridable methods and javadoc

MEZk opened the pull request:
Issue #3102 

>*no validation for overridable methods usage in class
>*no validation of javadoc where user need to put explanation of how to override it correctly.

1) Done. Now if method is public or protected and non-final (designed for extension) it has to have a javadoc.
2) Activated the check against Checkstyle source codes. Added suppression for *Test files as we do not need to add javadoc for UTs or make them final.

3) Question: should we check whether a method has @Override annotation? For example, if a method has @Override annotation, the Check will not rise violation. Rationale: method which is designed for extension has to have javadoc commend and class which extends the base class and overrides the method also inherits javadoc. And  if a method does not have @Override annotation, the Check will rise violation. In Checkstyle we have a lot of such cases, where there is @Override on public method, for example, check CIs violations.

What should be done in next PRs:

4)  Add option to specify the explanation text which should be included in javadoc to allow the method to be designed for extension. For example, “This implementation...”

P.S.
Reports are on the way ...

MEZk commented:
@romani 
>Please create report only for checkstyle project only

Done.
Diff report: http://mezk.github.io/i3102-DesignForExtension/diff_reports/index.html
Report after fix:  http://mezk.github.io/i3102-DesignForExtension/after/checkstyle.html
Excluded directories:
```
**/checkstyle-with-excludes/src/test/resources/**/*
**/checkstyle-with-excludes/src/it/resources/**/*
**/resources-noncompilable/**/*
```

As I sad the Check should also check whether a method has @Override annotation. In addition, I think it is reasonable to skip methods which have @Test annotation. What do you think?

romani commented:
Yes, we need additional property to ignore methods that are marked by some annotation. Default value is `@Override`. In our config we ahould suppress Test too

MEZk commented:
@romani 
1) Added additional option to ignore methods which have certain annotations. Default value was set to 'Override'.
2) Implemented new UT for the option.
3) Extended javadoc/xdoc.
4) Enforced the check against Checkstyle source codes. Fixed violations.
5) Performed refactoring of DesignForExtensionCheck in order to satisfy Checkstyle's checks and increase readability. In addition, I shortened the distance between methods.
6) Reports:

configuration of the check before fix:
```xml
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
        <module name="DesignForExtension"/>
    </module>
</module>
```

configuration of the check after fix:
```xml
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
        <module name="DesignForExtension">
            <property name="ignoredAnnotations" value="Override, Test, Before, After, BeforeClass, AfterClass"/>
        </module>
    </module>
</module>
```

[guava-diff-report](http://mezk.github.io/i3102-DesignForExtension/diff_reports/guava/index.html)

Questions:
1) Should we treat block comments before the method as javadoc comments? If yes, we need to add additional method into FileContents similar to getJavadocBefore, but which will return block comments. In the FileContents block comments are stored in clangComments. So the method will be named as getCLangCommentBefore or getBlockCommentBefore.

```java
/*
 *
 */
public int foo() {return 1;} // not a violation?
```

2) Should we treat single line comments themselves or single line comments arranged in block as javadoc comments? If yes, we also need to add additional method into FileContents similar to getJavadocBefore.

```java
// This implementation...
//
public int foo() {return 1;} // not a violation?

// This implementation...
public int foo2() {return 1;} // not a violation?
```

romani commented:
> Question: should we check whether a method has @Override annotation?

Previously I answered YES, but after some thought.....
We do validation of certain class. If method is public , does not matter any annotation, this method could be overridden. So its javadoc (of this certain method) has to mention about this. Override annotation tell that this method is inherited from another class, we have no access to another class, so it does not give us any answer on design on current class. It might be that current class is not designed for extension even its parent is designed for this.

One more point, we can not distinguish methods from Interface and abstract class. All such methods are marked with annotation. If class implements interface it is kind of designed for extension.


So to keep Override annotation by default is very questionable.


..... codereview is to be continued

romani commented:
@MEZk , CIs are read, please fix

MEZk commented:
@romani 
Postponed till #3466

MEZk commented:
@romani

>So to keep Override annotation by default is very questionable.

Excluded Override annotation from default values of ignoredAnnotations set. Only Test, Before, After, BeforeClass, AfterClass annotations are ignored by default now. For Checkstyle project the set of ignored annotations: Override, Test, Before, After, BeforeClass, AfterClass.

>1) Should we treat block comments before the method as javadoc comments? 
Code should have violation, as block comment is not javadoc, "*" identificator is required. 

Done. Now the check validates a presence of both ordinary javadoc (/** */) and block (/* */) comments. 
I implemented tree traverse algorith in order to check whether a method has empty body. Due to the fact that the algorith checks boundary conditions itself I did not break the logic apart and did not extract some pieces into seperate methods so as not to decriese readibility. CyclomaticComplexity of hasEmptyImplementation is 12 and I had to suppress it with a comment.

>2) Should we treat single line comments themselves or single line comments arranged in block as javadoc comments?
No, please read http://checkstyle.sourceforge.net/writingjavadocchecks.html#What_is_Javadoc_comment
Both cases should result in violation.

Done.

MEZk commented:
1) 

Replace 

```Even in library projects this Check most likely find classes that are not required to check.```

with

```Even in library projects this Check most likely find classes that are designed for extension by somebody.```


2) 
Replaced

```User need to use suppressions extensively to got a benefit from this Check and avoid false positives.```

with

```User need to use suppressions extensively to got a benefit from this Check , and keep in suppression all confirmed/known classes that are deigned for inheritance intentionally to let Check catch only new classes and bring this to team/user attention.```


I also rephrased javadoc/xdoc a bit and fixed grammar mistakes.

>guava-diff report
Method 'getTesters' is not designed for extension - needs to be abstract, final or empty.
We can not make such statement to user.

I will send you examples of new message.

3)
>Not clear why you did not subscribe to CLASS token.

Add a comment for CLASS_DEF as we discussed.

4)
>this variable could be calculated a bit later. please refactor following IF block.

Done.

5) 
>but you check for Static too, please update the name

Performed refactoring and remove isPrivateOrFinalOrAbstract method. Moved its content to canBeOverridden method.

6) 
>it is very hard to read this ..... user does not must ....... we need to rephrase this and state that we find classes that have overridable method (public or protected methods that are non-static, not-final, non-abstract) and have non-empty implementation.

Done. Extended xdoc/javadoc.

MEZk commented:
**DIFF REPORTS

Config:

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">


<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
        <module name="DesignForExtension"/>
    </module>
</module>

```
[sevntu-checkstyle-with-excludes](http://mezk.github.io/i3102-DesignForExtension/sevntu-checkstyle-with-excludes/)
[guava-mvnstyle](http://mezk.github.io/i3102-DesignForExtension/guava-mvnstyle/)
[pmd](http://mezk.github.io/i3102-DesignForExtension/pmd/)
[findbugs-with-excludes](http://mezk.github.io/i3102-DesignForExtension/findbugs-with-excludes/)
[spring](http://mezk.github.io/i3102-DesignForExtension/spring/)
[hibernate-orm-with-excludes](http://mezk.github.io/i3102-DesignForExtension/hibernate-with-excludes/)

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:
```diff
@@ -93,46 +126,169 @@
+        if (canBeOverridden(ast)
+                && !hasEmptyImplementation(ast)
+                && !hasJavadocComment(ast)
+                && !hasIgnoredAnnotation(ast, ignoredAnnotations)) {
```
please sort that expression by complexity of execution for performance reasons.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:
```diff
@@ -93,46 +126,169 @@
+
+            final DetailAST containerDef = getNearestClassOrEnumDefinition(ast);
+            if (canBeSubclassed(containerDef)
+                    && hasDefaultOrExplicitNonPrivateCtor(containerDef)) {
```
I think hasDefaultOrExplicitNonPrivateCtor should be inside canBeSubclassed.  Am I miss smth ?

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:
```diff
@@ -93,46 +126,169 @@
+
+    /**
+     * Checks whether a method has an empty implementation.
+     * Method is OK if its implementation is empty.
```
this line is useless. Please add note that you actually validate comments in method.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:
```diff
@@ -93,46 +126,169 @@
+        boolean hasEmptyBody = true;
+
+        final DetailAST implementation = ast.findFirstToken(TokenTypes.SLIST);
+        final boolean isNativeMethod = implementation == null;
```
native methods in java are marked by special modifier. native method != empty method (looks like , but not the same). Please use checking by modifier.
https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.4.3.4

romani commented:
0) I do not see changes in messages. Do you want to make them as separate commit ?

1) in reports there only removal of violations , that kind of expected, but this not show testing ...
Please generate report on checkstyle code before your fixed missed comments. 

2) Please split Check changes and modifications to different commits.

3) There is no description of Javadoc role for Check. Please add. User need to know all details.

4) please add to javadoc brief description that violation could be skipped by annotations. Please explain a reason why `Override` is not in defaults.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:
```diff
@@ -93,46 +126,169 @@
+     * @return true if a method has an empty implementation.
+     */
+    //-@cs[CyclomaticComplexity] We have to keep the complex logic
+    // as the method implements tree traverse algorithm.
```
bad excuse.
1) you can move check for native outside of this method.
2) please look at https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L495 for traversing a tree. That method is not complicated.
You intend is really to find first node that match predicate "!TokenUtils.isCommentType(currentToken.getType())" please make such function in TokenUtils.java .

MEZk commented:
0) Done in separate commit.
1) New reports will be posted a bit later.
2) Done.
3) Done.
4) Done.

>please sort that expression by complexity of execution for performance reasons.

Done.

>I think hasDefaultOrExplicitNonPrivateCtor should be inside canBeSubclassed.

Done

>this line is useless. Please add note that you actually validate comments in method.

> you can move check for native outside of this method

Done

>please look at https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L495 for traversing a tree. That method is not complicated.

Done

>You intend is really to find first node that match predicate "!TokenUtils.isCommentType(currentToken.getType())" please make such function in TokenUtils.java 

Done

>native methods in java are marked by special modifier. native method != empty method (looks like , but not the same). Please use checking by modifier.

Done

MEZk commented:
@romani 
Once again we have broken UTs due to internationalization problem. See Travis CI build log.

romani commented:
Ci is green ,
1)please make report over guava
2) please remove severity ignore from check at stress testing configs
https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/checks-nonjavadoc-error.xml , please review other files

MEZk commented:
@romani 

2) Done. https://github.com/checkstyle/contribution.git

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:
```diff
@@ -93,52 +132,169 @@
-                log(ast.getLineNo(), ast.getColumnNo(), MSG_KEY, name);
+        if (!hasJavadocComment(ast)
+                && (isNativeMethod(ast)
+                || !hasEmptyImplementation(ast))
```
indent this line to the right, to be visiable that is not on the same level

romani commented:
As you provide report , I will merge PR, all is ready. Great job !

MEZk commented:
@romani 

>1)please make report over guava

Done. See https://mezk.github.io/i3102-DesignForExtension/guava-mvnstyle/site/checkstyle.html
Config:
```xml
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
		<module name="DesignForExtension">
		  <property name="ignoredAnnotations" value="Test, Before, After, BeforeClass, AfterClass"/>
		</module>
    </module>
</module>
```

>2) Please generate report on checkstyle code before your fixed missed comments. 

Done. See https://mezk.github.io/i3102-DesignForExtension/checkstyle-master/site/checkstyle.html
Config:
```xml
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
		<module name="DesignForExtension">
		  <property name="ignoredAnnotations" value="Override, Test, Before, After, BeforeClass, AfterClass"/>
		</module>
    </module>
</module>
```
